core/lib/zksync_core/src/state_keeper/keeper.rs

```diff
@@ -136,25 +139,50 @@ impl ZkSyncStateKeeper {
 
         let previous_batch_protocol_version =
             self.io.load_previous_batch_version_id().await.unwrap();
+
+        let first_batch_in_shared_bridge = l1_batch_env.number == L1BatchNumber(1)
+            && protocol_version >= ProtocolVersionId::Version21;
         let version_changed = protocol_version != previous_batch_protocol_version;
 
-        let mut protocol_upgrade_tx = if pending_miniblocks.is_empty() && version_changed {
-            self.io.load_upgrade_tx(protocol_version).await
-        } else if !pending_miniblocks.is_empty() && version_changed {
-            // Sanity check: if `txs_to_reexecute` is not empty and upgrade tx is present for this block
-            // then it must be the first one in `txs_to_reexecute`.
-            if self.io.load_upgrade_tx(protocol_version).await.is_some() {
-                let first_tx_to_reexecute = &pending_miniblocks[0].txs[0];
-                assert_eq!(
-                    first_tx_to_reexecute.tx_format(),
-                    TransactionType::ProtocolUpgradeTransaction
-                )
+        let mut protocol_upgrade_tx = self.io.load_upgrade_tx(protocol_version).await;
```

Comment:
we're adding too much logic to this 'run_inner' function.

I'd like us to:
* move this logic to a separate method
* add unittests for this method.